### PR TITLE
[RFR] [BC Break] Update REST response format

### PR DESCRIPTION
**Pull request by fzaninotto**
This changes the expected return format of the `restClient`. Now it must systematically return an object with a `data` property.

```diff
 Type                 | Response format
 -------------------- | ----------------
 `GET_LIST`           | `{ data: {Record[]}, total: {int} }`
-`GET_ONE`            | `{Record}`
-`CREATE`             | `{Record}`
-`UPDATE`             | `{Record}`
-`DELETE`             | `{Record}`
-`GET_MANY`           | `{Record[]}`
-`GET_MANY_REFERENCE` | `{Record[]}`
+`GET_ONE`            | `{ data: {Record} }`
+`CREATE`             | `{ data: {Record} }`
+`UPDATE`             | `{ data: {Record} }`
+`DELETE`             | `{ data: {Record} }`
+`GET_MANY`           | `{ data: {Record[]} }`
+`GET_MANY_REFERENCE` | `{ data: {Record[]}, total: {int} }`
```

This removes a potential surprise for developers trying to build their own rest client.

Caution: This breaks all existing REST clients. Once merged, I'll notify the third-party rest client authors to help them migrate.

Refs #347

### CODE REVIEW

**Comment by mantis** [comment]
in your simple rest client, what are you expecting to be returned on a CREATE now - just an ID?

I ended up needing to modify create to be the same as default - i.e. case CREATE:                return {data: json};

**Comment by fzaninotto** [comment]
This is documented in https://marmelab.com/admin-on-rest/RestClients.html#writing-your-own-rest-client

**Comment by mantis** [comment]
let me rephrase that question a bit:

If the rest api is expected to return a complete object now - shouldn't this be using json.* and not params.data ?

         case CREATE:
 -            return { ...params.data, id: json.id };
 +            return { data: { ...params.data, id: json.id } };

**Comment by fzaninotto** [comment]
Not sure I understand you. 

```js
case CREATE:
     return { data: { ...params.data, id: json.id } };
```

Are you suggesting it returns instead:

```js
case CREATE:
     return { data: json };
```

This requires that the API returns the entire created object. Currently, we require only that it requires an id.

**Comment by mantis** [comment]
Initially, something was going wrong - I think i may have been returning [{ 'id': 1 }] as opposed to { 'id' :1 } or vice versa. 

However you are correct -  I am suggesting that CREATE should return the entire object. That would strike me as a more sensible default:

1) I'm not sure if that would be standard REST behaviour, but it would mirror how github api appears to work - https://developer.github.com/v3/projects/#create-a-repository-project - that would appear to return a complete object for instance

2) Let's suppose someone has a User - and asks for a username, id, url. If the client sends off url as "http://www.github.com " (note space at end) - it could be that the server would choose to trim() this string. At that point, in the current setup of returning just an ID - the client side and server side would be out of sync. Whereas if the default is to always return the entire created object, this would keep the two in sync automatically.

**Comment by fzaninotto** [comment]
That's debatable. It makes sense in your case, but it may break admin-on-rest for APIs that simply return an id.

I'd like feedback from more users before making a decision here.

**Comment by mantis** [comment]
@fzaninotto actually, as people can define their own REST client (that's what i did initially) and I expect a lot of people will do doing that - I guess the question is what the default should be. Both versions of the code obviously should work - I was struggling with the original, but I think that might have been because I was returning a single item array, not a single item. So probably nothing to do here - although for example use case above - I do wonder if that is a better model to encourage people to develop around. Anyway, I won't comment here anymore :)
